spec/index.adoc

```diff
@@ -4015,30 +4071,120 @@ Conditions::
   lookup(["canister",Q.canister_id,"certified_data"], Cert) = Found S.certified_data[Q.canister_id]
   lookup(["time"], Cert) = Found S.system_time // or “recent enough”
   Env = {
-    time = S.time[Q.receiver];
-    global_timer = S.global_timer[Q.receiver];
+    time = S.time[Q.canister_id];
+    global_timer = S.global_timer[Q.canister_id];
     balance = S.balances[Q.canister_id];
     freezing_limit = freezing_limit(S, Q.canister_id);
     certificate = Cert;
-    status = simple_status(S.canister_status[Q.receiver]);
-    canister_version = S.canister_version[Q.receiver];
+    status = simple_status(S.canister_status[Q.canister_id]);
+    canister_version = S.canister_version[Q.canister_id];
   }
 ....
 Read response::
-* If `F(Q.Arg, Q.sender, Env) = Trap trap` then
+* If `F(Q.Arg, Q.sender, Env)(S.canisters[Q.canister_id].wasm_state) = Trap trap` or `F(Q.Arg, Q.sender, Env)(S.canisters[Q.canister_id].wasm_state) = Return {response = NoResponse; …}` then
 +
 ....
 {status: rejected; reject_code: CANISTER_ERROR, reject_message: <implementation-specific>, error_code: <implementation-specific>}
 ....
-* Else if `F(Q.Arg, Q.sender, Env) = Return {response = Reject (code, msg); …}` then
+* Else if `F(Q.Arg, Q.sender, Env)(S.canisters[Q.canister_id].wasm_state) = Return {response = Reject (code, msg); …}` then `code = CANISTER_REJECT` and
++
+....
+{status: rejected; reject_code: CANISTER_REJECT: reject_message: <msg>, error_code: <implementation-specific>}
+....
+* Else if `F(Q.Arg, Q.sender, Env)(S.canisters[Q.canister_id].wasm_state) = Return {response = Reply R; …}` then
++
+....
+{status: success; reply: { arg : R } }
+....
+
+==== Composite query call
+
+Canister composite query calls to `/api/v2/canister/<ECID>/composite_query` can be executed directly. They can only be executed against canisters which have a status of `Running` and are also not frozen.
+
+During the execution of a composite query call, a certificate is provided to the canister that is valid, contains a current state tree (or “recent enough”; the specification is currently vague about how old the certificate may be), and reveals the canister’s <<system-api-certified-data,Certified Data>>.
+
+We define an auxiliary method that handles calls from composite query methods by performing a DFS call graph traversal.
+....
+composite_query_helper(S, Cycles, Caller, Canister_id, Method_name, Arg) =
```

Comment:
Do we need to track a stack of canisters called before to prevent reentrancy? Or is reentrancy allowed after all? @ulan

Reply:
Yes, reentrancy is allowed and we converged to the design where a query that's reentering the canister gets a fresh copy of the canister state. In other words, it doesn't see the changes done by previous queries of the current query call graph.

Reply:
Thank you for confirming! This means that nothing has to be changed here.